### PR TITLE
[JIT] Alias Doc Improvements

### PR DESCRIPTION
**Pull request by eellison**
Expand to cover type based alias analysis & handling of contained elements. cc @tlemo  @ansley

### CODE REVIEW

File: torch/csrc/jit/OVERVIEW.md [hunk 1]
```diff
@@ -1280,7 +1280,7 @@ This annotation language is consumed by the `FunctionSchema` parser, which produ
 [ir/alias_analysis.h](ir/alias_analysis.h)
 An alias analysis pass consumes the per-operator aliasing information to construct a database of aliasing and mutation relationships in a graph, called `AliasDb`. This section focuses on the alias analysis pass; the public interface to `AliasDb` will be described later.
 
-The core data structure in the AliasDb is called `AliasTracker`, which is a DAG where the edges are "may point to" relationships and the  vertices are aliasing `Element`s. The most common kind of `Element` is an IR `Value`, but there are other kinds of things that can alias that aren't first-class `Value`s in the IR, like wildcards or contained types (such as in a list or tuple).
+The core data structure in the AliasDb is called `AliasTracker`, which is a DAG where the edges are "may point to" relationships and the vertices are aliasing `Element`s. The most common kind of `Element` is an IR `Value`, but there are other kinds of things that can alias that aren't first-class `Value`s in the IR, like wildcards or contained types (such as in a list or tuple).
```
Comment (ansley) [inline torch/csrc/jit/OVERVIEW.md:1283]: Nit: I think `AliasTracker` has been renamed to `MemoryDAG`

File: torch/csrc/jit/ir/alias_analysis.cpp [hunk 2]
```diff
@@ -442,18 +442,17 @@ std::string AliasDb::toGraphviz() const {
...
+      dot << "  " << name(element) << " -> "
+          << name(memoryDAG_->fromIndex(pointedTo)) << "\n";
     }
-    if (!element->containedElements.empty()) {
-      for (const auto contained : element->containedElements) {
-        dot << "  " << name(element) << " -> "
-            << name(memoryDAG_->fromIndex(contained))
-            << " [style=dashed, color=blue]\n";
-      }
+    if (!element->pointsTo.count()) {
```
Comment (tlemo) [inline torch/csrc/jit/ir/alias_analysis.cpp:449]: is this where the new `verbose` argument should be checked?

File: torch/csrc/jit/python/python_ir.cpp [hunk 3]
```diff
@@ -220,6 +220,13 @@ void initPythonIRBindings(PyObject* module_) {
             AliasDb db(std::move(g));
             db.dump();
           })
+      .def(
+          "dump_alias_db_graphviz",
+          [](std::shared_ptr<Graph> g, bool verbose) {
+            AliasDb db(std::move(g));
+            std::cout << db.toGraphviz(verbose);
+          },
+          py::arg("verbose") = true)
```
Comment (tlemo) [inline torch/csrc/jit/python/python_ir.cpp:229]: can you please add `py::arg("graph")` too?

File: torch/csrc/jit/python/python_ir.cpp [hunk 4]
```diff
@@ -220,6 +220,13 @@ void initPythonIRBindings(PyObject* module_) {
             AliasDb db(std::move(g));
             db.dump();
           })
+      .def(
```
Comment (tlemo) [inline torch/csrc/jit/python/python_ir.cpp:223]: thanks for adding this

**Review by tlemo — APPROVED** [review]
thanks for expanding the aliasing documentation!

File: torch/csrc/jit/OVERVIEW.md [hunk 5]
```diff
@@ -1296,27 +1296,71 @@ As a more involved example, the following TorchScript snippet:
...
 - `c` is a fresh tensor (i.e. it doesn't point to anything) since it was created using the pure operation `2 * b`.
+- `d` is a fresh tensor, but because it enters a container (`ret`), it conservatively enters the wildcard set.
+- `ret` contains the Tensor wildcard set, which is visualized as the blue dotted line.
 
-The last point demonstrates a key concept: *leaf elements uniquely describe memory locations*. Since a leaf element doesn't point to anything, the memory that backs it must have been freshly allocated by some op. Thus we can use leaf elements to represent disjoint memory locations.
+Because `c` is a fresh tensor that does not point to anything, it is a leaf element. *leaf elements uniquely describe memory locations*. Since a leaf element doesn't point to anything, the memory that backs it must have been freshly allocated by some op. Thus we can use leaf elements to represent disjoint memory locations.
```
Comment (suo) [inline torch/csrc/jit/OVERVIEW.md:1344]: nit: `*leaf elements…` should be capitalized.

**Review by suo — APPROVED** [review]
lgtm

File: torch/csrc/jit/OVERVIEW.md [hunk 6]
```diff
@@ -1296,27 +1296,71 @@ As a more involved example, the following TorchScript snippet:
...
+Will produce the following IR:
+```
+graph(%a.1 : Tensor,
+      %b.1 : Tensor):
+  %2 : int = prim::Constant[value=2]() # test/elias.py:6:6
+  %6 : int = prim::Constant[value=1]() # test/elias.py:7:10
+  %14 : int = prim::Constant[value=4]() # test/elias.py:9:16
+  %17 : int = prim::Constant[value=0]() # test/elias.py:10:10
+  %c.1 : Tensor = aten::mul(%b.1, %2) # <string>:3:9
+  %d.1 : Tensor = aten::add(%c.1, %6, %6) # test/elias.py:7:6
```
Comment (ansley) [inline torch/csrc/jit/OVERVIEW.md:1318]: Can you delete #test/elias.py on all of these?

File: torch/csrc/jit/OVERVIEW.md [hunk 7]
```diff
@@ -1296,27 +1296,71 @@ As a more involved example, the following TorchScript snippet:
...
+      -> (%r.2)
+  %ret.1 : Tensor[] = prim::ListConstruct(%d.1, %r)
+  return (%ret.1)
 ```
-Will produce a graph like this:
 
-![AliasTracker graph](https://github.com/pytorch/pytorch/blob/master/docs/source/_static/img/aliastracker_graph.png)
+And the following AliasDb (visualized with `AliasDb::toGraphViz`)
+
+![AliasTracker graph1](https://github.com/pytorch/pytorch/blob/master/docs/source/_static/img/aliastracker_graph1.png)
```
Comment (ansley) [inline torch/csrc/jit/OVERVIEW.md:1335]: I don't see the pic but this might be on me

Comment (eellison) [reply torch/csrc/jit/OVERVIEW.md:1335]: They're included in the PR, i dont think the links will work until the PR lands.

**Comment by eellison** [comment]
image


image

**Pull request closed without merging**